data/lua/wml-flow.lua

```diff
@@ -179,36 +179,50 @@ function wml_actions.foreach(cfg)
 	local array = wml.array_variables[array_name]
 	if #array == 0 then return end -- empty and scalars unwanted
 	local item_name = cfg.variable or "this_item"
-	local this_item <close> = utils.scoped_var(item_name) -- if this_item is already set
 	local i_name = cfg.index_var or "i"
-	local i <close> = utils.scoped_var(i_name) -- if i is already set
+
+	-- Some protection against unsupported modification of the array while iterating.
+	-- Changing the array will simply be ignored during the loop, and then overwritten
+	-- by the contents of "array" afterwards, this is simply a method of trying to add
+	-- an error message instead of silently ignoring those changes.
 	local array_length = wml.variables[array_name .. ".length"]
+	local check_for_modifications = true
+	if array_name:find("^" .. item_name) then
+		-- Disable the protection for WML such as array=this_item or array=this_item.abilities
+		check_for_modifications = false
```

Comment:
I don't quite understand this. How does skipping the array length test in this circumstance make a difference?

Reply:
The protection is checking that the code in the loop isn't trying to modify the array via the name that exists outside the loop. However, if the name that exists outside the inner loop is `this_item` and it's shadowed by a scoped `this_item` inside the loop then the shadowing itself triggers a false positive for the protection code.